system/lib/wasmfs/backends/node_backend.cpp

```diff
@@ -0,0 +1,173 @@
+    // TODO: read
+    return 0;
+  }
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    // TODO: write
+    return 0;
+  }
+
+  void flush() override {}
+};
+
+class NodeDirectory : public Directory {
+public:
+  // The path to this directory on the underlying Node file system.
+  std::string path;
+
+  NodeDirectory(mode_t mode, backend_t backend, std::string path)
+    : Directory(mode, backend), path(path) {}
+
+private:
+  std::shared_ptr<File> getChild(const std::string& name) override {
+    auto childPath = path + '/' + name;
+    static_assert(std::is_same_v<mode_t, unsigned int>);
+    // TODO: also retrieve and set ctime, atime, ino, etc.
+    mode_t mode;
+    int exists = _wasmfs_node_get_mode(childPath.c_str(), &mode);
+    if (!exists) {
+      return nullptr;
+    }
+    std::shared_ptr<File> child;
+    if (S_ISREG(mode)) {
+      child = std::make_shared<NodeFile>(mode, getBackend(), childPath);
+    } else if (S_ISDIR(mode)) {
+      child = std::make_shared<NodeDirectory>(mode, getBackend(), childPath);
+    } else if (S_ISLNK(mode)) {
+      // return std::make_shared<NodeSymlink>(mode, getBackend(), childPath);
+    } else {
+      // Unrecognized file kind not made visible to WasmFS.
+      return nullptr;
+    }
+    child->locked().setParent(shared_from_this());
+    return child;
+  }
+
+  bool removeChild(const std::string& name) override {
+    // TODO
+    return false;
+  }
+
+  std::shared_ptr<File> insertChild(const std::string& name,
+                                    std::shared_ptr<File> file) override {
+    auto childPath = path + '/' + name;
+    auto mode = file->locked().getMode();
+    if (file->is<DataFile>()) {
+      if (_wasmfs_node_insert_file(childPath.c_str(), mode)) {
+        return nullptr;
+      }
+      std::static_pointer_cast<NodeFile>(file)->path = childPath;
+      return file;
+    } else if (file->is<Directory>()) {
+      if (_wasmfs_node_insert_directory(childPath.c_str(), mode)) {
+        return nullptr;
+      }
+      std::static_pointer_cast<NodeDirectory>(file)->path = childPath;
+      return file;
+    } else if (file->is<Symlink>()) {
+      // fs.linkSync(target, name)
+      assert(false && "Symlinks not implemented");
+      return nullptr;
+    } else {
+      assert(false && "Unimplemented file kind");
+      return nullptr;
+    }
+    return nullptr;
+  }
+
+  std::string getName(std::shared_ptr<File> file) override {
+    // TODO
+    return "";
+  }
+
+  size_t getNumEntries() override {
+    // TODO: optimize this?
+    return getEntries().size();
+  }
+
+  std::vector<Directory::Entry> getEntries() override {
+    std::vector<Directory::Entry> entries;
+    int err = _wasmfs_node_readdir(path.c_str(), &entries);
+    // TODO: Make this fallible
+    assert(err == 0);
+    return entries;
+  }
+};
+
+class NodeBackend : public Backend {
+  std::string rootPath;
```

Comment:
Is this the relative root of the "mount point"?

(If so, perhaps another name would be better? Just because root has a specific meaning already. Unless that meaning is also used for mount points?)

Reply:
Yeah, this is the underlying Node FS path of the directories or files that are mount points for this backend. Will change the name and add a comment.